cmd/kubeadm/app/phases/certs/certlist.go

```diff
@@ -397,3 +416,58 @@ func setCommonNameToNodeName() configMutatorsFunc {
 		return nil
 	}
 }
+
+// LeafCertificates returns non-CA certificates from the supplied Certificates.
+func LeafCertificates(c Certificates) (Certificates, error) {
+	certTree, err := c.AsMap().CertTree()
+	if err != nil {
+		return nil, err
+	}
+
+	var out Certificates
+	for _, leafCertificates := range certTree {
+		out = append(out, leafCertificates...)
+	}
+	return out, nil
+}
+
+type keyAndCSRCreator struct {
+	kubeadmConfig *kubeadmapi.InitConfiguration
+}
+
+func (o *keyAndCSRCreator) create(cert *KubeadmCert) error {
+	if o == nil {
+		return fmt.Errorf("%w: create was called on a nil pointer", errInvalid)
+	}
+	if cert == nil {
+		return fmt.Errorf("%w: cert was nil", errInvalid)
+	}
+	certDir := o.kubeadmConfig.CertificatesDir
+	name := cert.BaseName
+	if pkiutil.CSROrKeyExist(certDir, name) {
+		return fmt.Errorf("%w: key or CSR %s/%s", os.ErrExist, certDir, name)
+	}
+	cfg, err := cert.GetConfig(o.kubeadmConfig)
+	if err != nil {
+		return err
+	}
+	csr, key, err := pkiutil.NewCSRAndKey(cfg)
+	if err != nil {
+		return err
+	}
+	err = pkiutil.WriteKey(certDir, name, key)
+	if err != nil {
+		return err
+	}
+	err = pkiutil.WriteCSR(certDir, name, csr)
+	if err != nil {
+		return err
+	}
+	return nil
+}
```

Comment:
This where I can test the creation of a single key and CSR file and *some* of the error cases.
This could be made more generic  and reusable in future. See https://github.com/kubernetes/kubernetes/compare/master...wallrj:kubeconfig-builder